### PR TITLE
Add least common multiple function for UInt128

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: src/main/java/com/radixdlt/utils/MathUtils.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	 *
+	 * numbers must be non-null and non-zero. Otherwise, result is undefined.
+	 *
+	 * @param cap the cap to be used for the computation
+	 * @param numbers array of numbers of size atleast 1
+	 * @return null, if the least common multiple is greater than cap, otherwise the least common multiple
+	 * @throws ArrayIndexOutOfBoundsException if numbers is a zero length array
+	 * @throws NullPointerException if cap or numbers is null
+	 */
+	public static UInt128 cappedLCM(UInt128 cap, UInt128... numbers) {
```
Comment (flotothemoon) [inline src/main/java/com/radixdlt/utils/MathUtils.java:75]: It seems a bit awkward to have LCM in uppercase here and in lowercase above

File: src/main/java/com/radixdlt/utils/MathUtils.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		boolean overflow = !x.isZero() && !r.divide(x).equals(d);
+		return overflow ? null : r;
+	}
+
+	/**
+	 * Returns a capped least common multiple between an array of UInt128 numbers.
+	 * The cap acts as a ceiling. If the result exceeds cap, then this will return
+	 * null.
+	 *
+	 * numbers must be non-null and non-zero. Otherwise, result is undefined.
```
Comment (flotothemoon) [inline src/main/java/com/radixdlt/utils/MathUtils.java:67]: Could we check for non-null and non-zero to get rid of the "undefined" behaviour?

File: src/main/java/com/radixdlt/utils/MathUtils.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		throw new IllegalStateException("Cannot instantiate");
+	}
+
+	/**
+	 * Euclid's algorithm to find greatest common divisor
+	 * @param x first number
+	 * @param y second number
+	 * @return greatest common divisor between x and y
+	 */
+	private static UInt128 gcd(UInt128 x, UInt128 y) {
```
Comment (flotothemoon) [inline src/main/java/com/radixdlt/utils/MathUtils.java:21]: What is the reasoning for making this `private`?

File: src/main/java/com/radixdlt/utils/MathUtils.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		return (y.isZero()) ? x : gcd(y, x.remainder(y));
+	}
+
+	/**
+	 * Least common multiple computed via reduction by gcd
+	 * @param x first number
+	 * @param y second number
+	 * @return least common multiple between x and y
+	 */
+	private static UInt128 lcm(UInt128 x, UInt128 y) {
```
Comment (flotothemoon) [inline src/main/java/com/radixdlt/utils/MathUtils.java:41]: This would also be a useful `public` utility to expose.

**Review by flotothemoon — APPROVED** [review]
Good by me. It might be useful to also expose the new `gcd` and `lcm` methods.

File: src/main/java/com/radixdlt/utils/MathUtils.java (hunk 2 above)
Comment (talekhinezh) [reply src/main/java/com/radixdlt/utils/MathUtils.java:67]: I think keep low level math utils should be mostly branch free for speed.

File: src/main/java/com/radixdlt/utils/MathUtils.java (hunk 3 above)
Comment (talekhinezh) [reply src/main/java/com/radixdlt/utils/MathUtils.java:21]: Didn't want to write unit tests :)

**Pull request merged**